File: Core/src/main/java/com/inverce/mod/core/Lifecycle.java
package com.inverce.mod.core;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.RestrictTo;

import com.inverce.mod.core.interfaces.ActivityStateListener;
import com.inverce.mod.core.interfaces.LifecycleState;
import com.inverce.mod.core.internal.IMInternal;

/**
 * The Lifecycle utilities.
 */
public class Lifecycle {
    @NonNull
    private static LifecycleState currentLifecycleState = LifecycleState.NotCreated;
    private static ActivityStateListener listener;
    private static int activityHash;

    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    public static void initialize() {
        if (IMInternal.get().isInEdit()) return;
        IM.application().registerActivityLifecycleCallbacks(new StatesAdapterImpl());
    }

    /**
     * Sets ActivityState listener.
     *
     * @param listener the listener
     */
    public static void setListener(ActivityStateListener listener) {
        Lifecycle.listener = listener;
    }

    /**
     * Gets activity state.
     *
     * @return the activity state
     */
    @NonNull
    public static LifecycleState getActivityState() {
        return currentLifecycleState;
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    synchronized static void onActivityState(@NonNull LifecycleState state, @NonNull Activity activity, Bundle extra) {
        if (state.ordinal() > 3 && activity.hashCode() != activityHash) {
            // don't post activity state change for previous activity
            return;
        }

        currentLifecycleState = state;

        if (state.ordinal() < 4) {
            activityHash = activity.hashCode();
            IMInternal.get().setActivity(activity);
        }

        if (listener != null) {
            listener.activityStateChanged(state, activity, extra);
        }
    }

    private static class StatesAdapterImpl implements Application.ActivityLifecycleCallbacks {
        public void onActivityCreated(@NonNull Activity activity, Bundle savedInstanceState) {
            onActivityState(LifecycleState.Created, activity, savedInstanceState);
        }

        public void onActivityStarted(@NonNull Activity activity) {
            onActivityState(LifecycleState.Started, activity, null);
        }

        public void onActivityResumed(@NonNull Activity activity) {
            onActivityState(LifecycleState.Resumed, activity, null);
        }

        public void onActivityPaused(@NonNull Activity activity) {
            onActivityState(LifecycleState.Paused, activity, null);
        }

        public void onActivityStopped(@NonNull Activity activity) {
            onActivityState(LifecycleState.Stopped, activity, null);
        }

        public void onActivitySaveInstanceState(@NonNull Activity activity, Bundle outState) {
            onActivityState(LifecycleState.SaveInstanceState, activity, outState);
        }

        public void onActivityDestroyed(@NonNull Activity activity) {
            onActivityState(LifecycleState.Destroyed, activity, null);
        }
    }
}

File: Core/src/main/java/com/inverce/mod/v2/core/functional/Lifecycle.kt
package com.inverce.mod.v2.core.functional

import android.app.Activity
import android.app.Application
import android.os.Bundle
import android.support.annotation.RestrictTo
import com.inverce.mod.v2.core.application
import com.inverce.mod.v2.core.internal.IMInternal
import java.lang.ref.WeakReference

/**
 * The Lifecycle utilities.
 */
object Lifecycle {
    /**
     * Gets activity state.
     *
     * @return the activity state
     */
    var activityState = LifecycleState.NotCreated
        private set

    private var listener: ActivityStateListener? = null
    private var activityHash: Int = 0

    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    fun initialize() {
        if (IMInternal.isInEdit) return
        application.registerActivityLifecycleCallbacks(StatesAdapterImpl())
    }

    /**
     * Sets ActivityState listener.
     *
     * @param listener the listener
     */
    fun setListener(listener: ActivityStateListener) {
        Lifecycle.listener = listener
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    @Synchronized
    internal fun onActivityState(state: LifecycleState, activity: Activity, extra: Bundle?) {
        if (state.ordinal > 3 && activity.hashCode() != activityHash) {
            // don't post activity state change for previous activity
            return
        }

        activityState = state

        if (state.ordinal < 4) {
            activityHash = activity.hashCode()
            IMInternal.activity = WeakReference(activity)
        }


        listener?.activityStateChanged(state, activity, extra)
    }


}

class StatesAdapterImpl : Application.ActivityLifecycleCallbacks {
    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle) {
        Lifecycle.onActivityState(LifecycleState.Created, activity, savedInstanceState)
    }

    override fun onActivityStarted(activity: Activity) {
        Lifecycle.onActivityState(LifecycleState.Started, activity, null)
    }

    override fun onActivityResumed(activity: Activity) {
        Lifecycle.onActivityState(LifecycleState.Resumed, activity, null)
    }

    override fun onActivityPaused(activity: Activity) {
        Lifecycle.onActivityState(LifecycleState.Paused, activity, null)
    }

    override fun onActivityStopped(activity: Activity) {
        Lifecycle.onActivityState(LifecycleState.Stopped, activity, null)
    }

    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
        Lifecycle.onActivityState(LifecycleState.SaveInstanceState, activity, outState)
    }

    override fun onActivityDestroyed(activity: Activity) {
        Lifecycle.onActivityState(LifecycleState.Destroyed, activity, null)
    }
}

enum class LifecycleState {
    NotCreated, Created, Started, Resumed, Paused, Stopped, SaveInstanceState, Destroyed
}

interface ActivityStateListener {
    fun activityStateChanged(state: LifecycleState, activity: Activity, extra: Bundle?)
}